synapse/storage/database.py

```diff
@@ -1291,14 +1275,15 @@ def simple_upsert_txn_emulated(
             insertion_values: additional key/values to use only when inserting
             where_clause: An index predicate to apply to the upsert.
             lock: True to lock the table when doing the upsert.
+                Must not be False unless the table has already been locked.
         Returns:
             Returns True if a row was inserted or updated (i.e. if `values` is
             not empty then this always returns True)
         """
         insertion_values = insertion_values or {}
 
-        # We need to lock the table :(, unless we're *really* careful
         if lock:
```

Comment:
Why are we keeping this one conditional?

Reply:
It's to support `simple_upsert_many_txn_emulated`, which does its own locking beforehand.
https://github.com/matrix-org/synapse/blob/9cae44f49e6bf4f6b8a20ab11a65da417bb1565f/synapse/storage/database.py#L1499-L1509